src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -27,178 +25,78 @@ public class CompilerCache : ICompilerCache
         /// Initializes a new instance of <see cref="CompilerCache"/> populated with precompiled views
         /// discovered using <paramref name="provider"/>.
         /// </summary>
-        /// <param name="assemblyProvider">The <see cref="IAssemblyProvider"/> that provides assemblies
-        /// for precompiled view discovery.</param>
+        /// <param name="razorFileInfoCollections">The sequence of <see cref="RazorFileInfoCollection"/> that provides
+        /// information for precompiled view discovery.</param>
         /// <param name="loaderContextAccessor">The <see cref="IAssemblyLoadContextAccessor"/>.</param>
         /// <param name="optionsAccessor">An accessor to the <see cref="RazorViewEngineOptions"/>.</param>
-        public CompilerCache(IAssemblyProvider assemblyProvider,
-                             IAssemblyLoadContextAccessor loadContextAccessor,
-                             IOptions<RazorViewEngineOptions> optionsAccessor)
-            : this(GetFileInfos(assemblyProvider.CandidateAssemblies),
-                  loadContextAccessor.GetLoadContext(RazorFileInfoCollectionType.Assembly),
+        public CompilerCache(
+            IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
+            IAssemblyLoadContextAccessor loadContextAccessor,
+            IOptions<RazorViewEngineOptions> optionsAccessor)
+            : this(razorFileInfoCollections,
+                  loadContextAccessor.GetLoadContext(typeof(CompilerCache).GetTypeInfo().Assembly),
                   optionsAccessor.Options.FileProvider)
         {
         }
 
-        internal CompilerCache(IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
-                               IAssemblyLoadContext loadContext,
-                               IFileProvider fileProvider)
+        internal CompilerCache(
+            IEnumerable<RazorFileInfoCollection> razorFileInfoCollections,
+            IAssemblyLoadContext loadContext,
+            IFileProvider fileProvider)
```

Comment:
My memory is a little bit fuzzy about that conversation, but the complexity of what I'm suggesting isn't very high at all.

```
public interface ICompileCacheProvider
{
    ICompilerCache Cache { get; } // This does the lazy-init
}
```

It's easy because inside the implementation, cache is just null or initialized. I don't feel very strongly about this, it just seems like a much more friendly way to do this in combination with DI.